src/Castle.Core/DynamicProxy/ProxyUtil.cs

```diff
@@ -109,17 +150,134 @@ public static bool IsAccessible(MethodBase method, out string message)
 			return false;
 		}
 
+		/// <summary>
+		/// Checks whether the specified type is accessible to DynamicProxy.</summary>
+		/// <param name="type">The type to check.</param>
+		/// <returns><c>true</c> if the type is accessible to DynamicProxy, <c>false</c> otherwise.</returns>
+		public static bool IsAccessible(Type type)
+		{
+			return IsAccessibleType(type);
+		}
+
+		internal static bool IsAccessibleType(Type target)
+		{
+			var typeInfo = target.GetTypeInfo();
+
+			var isPublic = typeInfo.IsPublic || typeInfo.IsNestedPublic;
+			if (isPublic)
+			{
+				return true;
+			}
+
+			var isTargetNested = target.IsNested;
+			var isNestedAndInternal = isTargetNested && (typeInfo.IsNestedAssembly || typeInfo.IsNestedFamORAssem);
+			var isInternalNotNested = typeInfo.IsVisible == false && isTargetNested == false;
+			var isInternal = isInternalNotNested || isNestedAndInternal;
+			if (isInternal && AreInternalsVisibleToDynamicProxy(typeInfo.Assembly))
+			{
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		///   Checks whether the specified method is accessible to DynamicProxy.
+		///   Unlike with <see cref="IsAccessible(MethodBase)"/>, the declaring type's accessibility is ignored.
+		/// </summary>
+		/// <param name = "method">The method to check.</param>
+		/// <returns><c>true</c> if the method is accessible to DynamicProxy, <c>false</c> otherwise.</returns>
+		internal static bool IsAccessibleMethod(MethodBase method)
+		{
+			// Accessibility supported by the full framework and CoreCLR
+			if (method.IsPublic || method.IsFamily || method.IsFamilyOrAssembly)
+			{
+				return true;
+			}
+
+			if (method.IsFamilyAndAssembly)
```

Comment:
This could be merged with the previous if statement, actually `method.IsFamilyAndAssembly == true` means the method is accessible to derived classes inside this assembly, meaning we should actually be performing the same `AreInternalsVisibleToDynamicProxy` check below with these.

I guess this hasn't been a problem because I don't think you can define a method like this in C#, `protected internal` gives you `IsFamilyOrAssembly`.